### ISSUE TITLE
Type demotion in `jax.jacobian`

### ISSUE THREAD

**Issue by romanngg** [google/jax#6638]
Example:
```python
from jax import jacobian
from jax.config import config
import jax.numpy as jnp
config.update('jax_enable_x64', True)

def f(x, y):
  return x * y

x = jnp.ones((), jnp.float32)
y = jnp.ones((), jnp.float64)

f(x, y).dtype, jacobian(f)(x, y).dtype
```
Gives
```python
(dtype('float64'), dtype('float32'))
```

However, `df/dx = y` should have type `jnp.float64`.

(this seems to cause a type mismatch in https://github.com/google/neural-tangents/issues/112)

**Comment by jakevdp** [comment]
Thanks - this is a tricky one, because the backward pass by design produces a result with the same dtype as the input; here the default arguments ensure this is the same as `jacobian(f, argnums=0)(x, y)`, and so the dtype will be the same as that of `x`.

Note that forward-mode is different, so one workaround may be to use `jacfwd` rather than `jacobian`, which is aliased to `jacrev`.

Conceptually, the issue is that our jaxpr for `f` looks like this:
```python
>>> make_jaxpr(partial(f, y=y)(x)
{ lambda  ; a.
  let b = convert_element_type[ new_dtype=float64
                                weak_type=False ] a
      c = mul b 1.0
  in (c,) }
```
so naturally the backward pass finishes with a conversion to float32:
```
>>> primals, f_vjp = jax.vjp(partial(f, y=y), x)
>>> jax.make_jaxpr(f_vjp)(primals)
{ lambda  ; a.
  let b = mul a 1.0
      c = convert_element_type[ new_dtype=float32
                                weak_type=False ] b
  in (c,) }
```
Again this is by design: the transpose of a conversion from float32 to float64 is a conversion from float64 to float32.

Conceptually what we would want is to remove that conversion from the backward pass in cases that it makes sense... but I'm not certain how we would robustly determine when it makes sense.

I'd be open to any ideas folks might have.

**Comment by jakevdp** [comment]
Here's a repo that's slightly more to the point:
```python
from jax import jacobian
import jax.numpy as jnp

def f(x):
  return x.astype('float16')

x = jnp.array(1.0)
print(f(x).dtype, jacobian(f)(x).dtype)
# float16 float32
```

**Comment by froystig** [comment]
The observed behavior is correct, or at least consistent with our differentiation semantics. The cotangent of the first argument of `f` has type `float32`, since the first argument has type `float32`. The tangent of the output of `f` has type `float16`, since the output has type `float16`.

What's perhaps unexpected is that `jacobian` is an alias for `jacrev` rather than `jacfwd`, as @jakevdp notes. I'd go further and offer that calling _either_ function `jacobian` (via alias) is imprecise in this setting. Especially when a function has differing input/output types, it helps to take the more general view of its Jacobian (at a primal point) as a *function*.

Namely, for a fixed primal point `x`, the Jacobian `J(x)` is a (linear) function from input tangents to output tangents. Its transpose `J'(x)` is a function from output (co)tangents to input (co)tangents. In this view, `jacfwd` is the Jacobian map `J(x)` applied to the standard basis vectors of its domain, and `jacrev` is the Jacobian map's transpose `J'(x)` applied to the standard basis vectors of its domain.

So the outputs of `jacfwd` and `jacrev` are elements of different vector spaces, typed accordingly. Neither object exactly captures the full notion of "the Jacobian"—that's a function.

Perhaps we can improve documentation around `jacobian`, `jacfwd`, and `jacrev` to clarify their meaning.

**Comment by mattjj** [comment]
@froystig 's analysis is perfect!

I think when I first saw Roman mention this bug in chat, I had forgotten `jacobian` meant `jacrev` (and thought it might mean `jacfwd`), falling into exactly the trap that Roy described.

**Comment by romanngg** [comment]
Thank you for a great detailed explanation! This now makes sense to me.

**Comment by jakevdp** [comment]
Closing as this is working as intended.

**Comment by froystig** [comment]
There's still potentially something to do here:

> Perhaps we can improve documentation around `jacobian`, `jacfwd`, and `jacrev` to clarify their meaning.

I meant to indicate that with the label. Re-opening and will try to get around to this (or decide to close again).
